### PR TITLE
feat(jans-cedarling): Synchronize JWT bootstrap defaults & enforce strict validation as default 

### LINKED ISSUE #14097: feat(jans-cedarling): Synchronize JWT bootstrap defaults & enforce strict validation as default
# Synchronize JWT bootstrap defaults & enforce strict validation as default

## Background

`JwtConfig` defaults are reachable via two paths today:

1. **Bootstrap RAW path** (`BootstrapConfigRaw` → `JwtConfig`) — used by env vars / JSON / YAML config.
2. **Rust builder path** (`JwtConfig::default()`) — used by embedders constructing config directly in Rust.

These two paths produce **different** values, which is confusing for library users and inconsistent with the docs.

| Field | RAW default (env path) | `JwtConfig::default()` | Docs claim |
|---|---|---|---|
| `jwt_sig_validation` | `Disabled` (false) | `true` | `disabled` |
| `jwt_status_validation` | `Disabled` (false) | `true` | `disabled` |
| `token_cache_max_ttl_secs` | `0` | `300` (5 min) | `0` |

Additionally, the current "permissive by default" stance on the RAW path is a **security concern**: an empty bootstrap config accepts JWTs **without signature validation**, which makes spoofing trivial.

## Goals

1. **Single source of truth** — both paths must produce identical defaults.
2. **Strict-by-default** — security-critical validation flags ON by default. Fail-safe behavior: misconfigured Cedarling refuses tokens loudly instead of silently accepting them.
3. **Documentation matches code.**

## Proposed defaults (strict, identical on both paths)

| Field | New default | Rationale |
|---|---|---|
| `jwt_sig_validation` | **`true`** (Enabled) | Without signature validation, a JWT is just plain JSON. Critical. If no JWKS is configured, auth fails loudly — correct fail-safe behavior. |
| `jwt_status_validation` | **`true`** (Enabled) | Catches revoked tokens. `status_list_endpoint` is `Option<Url>`; absent endpoint → status check is naturally skipped (graceful fallback). |
| `signature_algorithms_supported` | all supported (unchanged) | `none` is not in the `Algorithm` enum. HS\* hardening can be a separate discussion. |
| `token_cache_max_ttl_secs` | **`5`** (seconds) | Compromise between perf (repeated requests with the same token use cached result) and freshness (revocation / status changes picked up within ~5s). Prevents unbounded cache growth for tokens with no `exp`. |
| `token_cache_capacity` | `100` (unchanged) | |
| `token_cache_earliest_expiration_eviction` | `true` (unchanged) | |
| `jwks_refresh_min_interval` | `30` (unchanged) | |

### 1. Extract shared constants

In `cedarling/src/bootstrap_config/jwt_config.rs`, expose:

```rust
impl JwtConfig {
    pub const DEFAULT_TOKEN_CACHE_CAPACITY: usize = 100;
    pub const DEFAULT_TOKEN_CACHE_MAX_TTL_SECS: usize = 5;
    pub const DEFAULT_JWKS_REFRESH_MIN_INTERVAL: u64 = 30;
}
```

### 2. Sync `JwtConfig::default()` to strict

Replace literal values in `impl Default for JwtConfig` to use the constants above and set:
- `jwt_sig_validation: true`
- `jwt_status_validation: true`
- `token_cache_max_ttl_secs: Self::DEFAULT_TOKEN_CACHE_MAX_TTL_SECS`

Update docstring to state: "Returns strict-by-default validation; identical to the values produced from an empty bootstrap configuration."

`new_without_validation()` stays — explicit opt-out constructor for tests/embedders that need permissive mode.

### 3. Flip RAW defaults to strict

In `cedarling/src/bootstrap_config/raw_config/default_values.rs`, add:

```rust
pub(super) fn default_enabled_feature_toggle() -> FeatureToggle {
    FeatureToggle::Enabled
}

pub(super) fn default_token_cache_max_ttl() -> usize {
    JwtConfig::DEFAULT_TOKEN_CACHE_MAX_TTL_SECS
}
```

In `cedarling/src/bootstrap_config/raw_config/config.rs`:
- `jwt_sig_validation` → `#[serde(default = "default_enabled_feature_toggle")]`
- `jwt_status_validation` → `#[serde(default = "default_enabled_feature_toggle")]`
- `token_cache_max_ttl` → `#[serde(default = "default_token_cache_max_ttl")]`

Switch `default_token_cache_capacity` and `default_jwks_refresh_min_interval` to use the new public constants (eliminates the circular `JwtConfig::default()` call from serde defaults).

### 4. Bug fix in token cache

`cedarling/src/jwt/token_cache.rs` — `save()` currently calls `SparKV::set()` (no TTL) when both the token has no `exp` and `max_ttl == 0`. With the underlying SparKV `max_ttl` config also being `0`, this contradicts the documented behavior ("token is not cached at all"). Fix: when `cache_duration` returns `None`, skip the write entirely. Update docstring of `TokenCache::new()` and `save()` to describe the three TTL paths precisely.

### 5. Tests

- Verify all existing tests still pass; expect adjustments where `..Default::default()` produced permissive config and the test relied on that implicitly.
- Update `default_config.yaml` if necessary so `test_load_default_config` assertions reflect the new defaults.
- Add a regression test: empty bootstrap config + token without valid signature → authorization is rejected.
- Add a regression test for token cache: `max_ttl=0` + token without `exp` → not cached (`find` returns `None` on second lookup).

### 6. Docs

Update `docs/cedarling/reference/cedarling-properties.md`:
- `CEDARLING_JWT_SIG_VALIDATION` — default `enabled`.
- `CEDARLING_JWT_STATUS_VALIDATION` — default `enabled`.
- `CEDARLING_TOKEN_CACHE_MAX_TTL` — default `5` (seconds). Explain trade-off (perf vs freshness) and that `0` disables the cap.
- Add a "Security" callout in the JWT validation reference noting that disabling these flags is **not recommended** outside of testing.

## Breaking change notice

This **is** a breaking change for existing deployments that relied on the implicit permissive defaults of an empty bootstrap config:

- Deployments without configured trusted issuers / `CEDARLING_LOCAL_JWKS` will start rejecting JWTs (which is the correct fail-safe behavior, but needs an explicit migration step).
- Token caching becomes shorter (5s vs effectively unlimited for tokens with long `exp`). Performance impact is negligible for typical workloads but should be called out.

Release note must explicitly tell operators to:
1. Configure JWKS source (local file or trusted issuers via OIDC discovery) before upgrading, OR
2. Explicitly opt out by setting `CEDARLING_JWT_SIG_VALIDATION=disabled` / `CEDARLING_JWT_STATUS_VALIDATION=disabled` (not recommended for production).

## Out of scope (potential follow-ups)

- Hardening `signature_algorithms_supported` to asymmetric-only by default (mitigates alg-confusion attacks involving HS\*).
- Reviewing other "Disabled by default" `FeatureToggle` fields for security implications (`CEDARLING_LOCK_ACCEPT_INVALID_CERTS` is already correctly `Disabled`).

## Acceptance criteria

- [ ] `BootstrapConfigRaw::default().try_into::<JwtConfig>()` produces a value equal to `JwtConfig::default()`.
- [ ] Both have `jwt_sig_validation == true`, `jwt_status_validation == true`, `token_cache_max_ttl_secs == 5`.
- [ ] Token without `exp` is not cached when `max_ttl == 0`.
- [ ] Docs in `cedarling-properties.md` match the actual code defaults.
- [ ] All existing tests pass; new regression tests added.

### PR DESCRIPTION
**Pull request by olehbozhok**
### Prepare

- [x] Read [PR guidelines](https://github.com/JanssenProject/jans/blob/main/docs/CONTRIBUTING.md#prs)
- [x] Read [license information](https://github.com/JanssenProject/jans/blob/main/LICENSE)

-------------------

#### Target issue
  
  
  
  
closes #14097

#### Implementation Details
  

-------------------
### Test and Document the changes
- [x] Static code analysis has been run locally and issues have been fixed
- [x] Relevant unit and integration tests have been added/updated
- [x] Relevant documentation has been updated if any (i.e. user guides, installation and configuration guides, technical design docs etc)

Please check the below before submitting your PR. The PR will not be merged if there are no commits that start with `docs:` to indicate documentation changes or if the below checklist is not selected.
- [ ] **I confirm that there is no impact on the docs due to the code changes in this PR.**

## Summary by CodeRabbit

* **Documentation**
  * Updated JWT configuration docs and default config sample to reflect new defaults and clarified token-cache semantics.

* **Behavior Changes**
  * JWT signature and status validation enabled by default.
  * Token cache default max TTL reduced to 5 seconds; effective TTL computation clarified.
  * Tokens are not cached when a positive TTL cannot be determined.
  * Missing status-list or OpenID info now skips status initialization and logs a warning.

[Review Change Stack](https://app.coderabbit.ai/change-stack/JanssenProject/jans/pull/14098?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by mo-auto** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/janssenproject/pr-checks/3808db45-c557-49b5-aaf7-0045dc67efd7?refs=3858ea85d0efad1fcc539da6e6b3b1192d5e7e36&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

**Comment by mo-auto** [comment]
Error: Hi @olehbozhok, You did not reference an open issue in your [PR](#14098). I attempted to create an [issue](#14099) for you.
Please update that [issues'](#14099) title and body and make sure I correctly referenced it in the above PRs body.

File: jans-cedarling/cedarling/src/bootstrap_config/jwt_config.rs [hunk 1]
```diff
@@ -68,27 +68,52 @@ pub(crate) const DEFAULT_JWKS_REFRESH_INTERVAL_SECS: u64 = 3600;
...
         let config = Self {
             jwks: None,
             jwt_sig_validation: true,
             jwt_status_validation: true,
             signature_algorithms_supported: HashSet::new(),
-            token_cache_capacity: 100,
+            token_cache_capacity: Self::DEFAULT_TOKEN_CACHE_CAPACITY,
             token_cache_earliest_expiration_eviction: true,
-            token_cache_max_ttl_secs: 60 * 5, // 5min
+            token_cache_max_ttl_secs: Self::DEFAULT_TOKEN_CACHE_MAX_TTL_SECS,
```
[the remark this answers is not in the document]
Comment (olehbozhok) [reply jans-cedarling/cedarling/src/bootstrap_config/jwt_config.rs:105]: It is OK, if the user needs, he can adjust. It is to avoid blowing the cache if we have a lot of tokens, and it allows us to periodically check if the token is still valid, maybe it was evicted.

File: jans-cedarling/cedarling/src/jwt/status_list/cache.rs [hunk 2]
```diff
@@ -59,14 +63,34 @@ impl StatusListCache {
...
+                            "issuer '{}' does not publish a status_list_endpoint; \
+                             status validation skipped",
+                            openid_config.issuer,
+                        ),
+                        Some(LogLevel::WARN),
+                    ));
+                }
+                return Ok(());
+            },
+        };
```
[the remark this answers is not in the document]
Comment (olehbozhok) [reply jans-cedarling/cedarling/src/jwt/status_list/cache.rs]: 1. The caller already ensures init_for_iss only runs when jwt_status_validation == true (line 224).
2. Adding a jwt_status_validation parameter to init_for_iss would be dead parameter pollution — it's always true when this function is called.
3. A hard error here would mean Cedarling can't start with jwt_status_validation=enabled if the issuer doesn't publish a status_list_endpoint. That's worse than silently skipping — operators who need status validation would configure their issuer to publish one; operators who can't would be blocked from upgrading. The WARN log is visible in their logging pipeline.

**Comment by olehbozhok** [comment]
@coderabbitai if you don't have other suggestions, can you approve pls?

**Review by haileyesus2433 — DISMISSED** [review]

File: jans-cedarling/cedarling/src/bootstrap_config/jwt_config.rs (hunk 1 above)
Comment (tareknaser) [inline jans-cedarling/cedarling/src/bootstrap_config/jwt_config.rs:105]: The new default (5 sec) is wired up here but the docstring on the `token_cache_max_ttl_secs` field itself is now out of date:
```rust
/// Maximum TTL (in seconds) for cached tokens.
/// Zero means no TTL limit is applied.
///
/// It is recommended to keep this value within a few minutes to prevent the
/// cache from growing excessively.
pub token_cache_max_ttl_secs: usize,
```

File: docs/cedarling/reference/cedarling-properties.md [hunk 3]
```diff
@@ -110,16 +110,16 @@ Also called Token-based Access Control (TBAC). This is the recommended authoriza
 
 **JWT and cryptographic behavior:**
 
-- **`CEDARLING_JWT_SIG_VALIDATION`** : `enabled` | `disabled` -- Whether to check the signature of all JWT tokens. When enabled, this requires the `iss` claim to be present in all tokens and the issuer URL must use the `https` scheme. Default is `disabled`.
-- **`CEDARLING_JWT_STATUS_VALIDATION`** : `enabled` | `disabled` -- Whether to check the status of the JWT. On startup, the Cedarling should fetch and retrieve the latest Status List JWT from the `.well-known/openid-configuration` via the `status_list_endpoint` claim and cache it. See the [IETF Draft](https://datatracker.ietf.org/doc/draft-ietf-oauth-status-list/) for more info. Default is `disabled`.
+- **`CEDARLING_JWT_SIG_VALIDATION`** : `enabled` | `disabled` -- Whether to check the signature of all JWT tokens. When enabled, this requires the `iss` claim to be present in all tokens and the issuer URL must use the `https` scheme. Default is `enabled`. **Disabling this is strongly discouraged outside of testing**: a JWT without signature validation is just plain JSON and is trivially spoofable.
```
Comment (tareknaser) [inline docs/cedarling/reference/cedarling-properties.md]: `jans-cedarling/cedarling/config/default_config.yaml` still has `CEDARLING_JWT_STATUS_VALIDATION: "disabled"` and `CEDARLING_TOKEN_CACHE_MAX_TTL` is not set. 

the bundled default sample shows the opposite of what is now the real default. Could you flip it to `"enabled"` (and either add or drop the `MAX_TTL` line so it matches `5`) so the sample matches the docs?

File: docs/cedarling/reference/cedarling-properties.md [hunk 4]
```diff
@@ -110,16 +110,16 @@ Also called Token-based Access Control (TBAC). This is the recommended authoriza
...
 
 **Token cache:**
 
-- **`CEDARLING_TOKEN_CACHE_MAX_TTL`** : Maximum token cache TTL in seconds. The token cache avoids decoding and validating the same token twice. Default is `0`, which disables the maximum TTL — in that case, the token's `exp` claim is used to compute the cache entry TTL. If the token has no `exp` claim and this is `0`, the token is not cached at all. If the token has no `exp` claim and this is > 0, this value is used as the cache TTL fallback.
+- **`CEDARLING_TOKEN_CACHE_MAX_TTL`** : Maximum token cache TTL in seconds. The token cache avoids decoding and validating the same token twice. Default is `5` seconds — small enough that revocation and status-list changes are picked up quickly, large enough to amortise repeated requests for the same token. Effective TTL is `min(time-until-exp, max_ttl)` when both apply. Setting this to `0` disables the cap — the token's `exp` claim is used as the entry TTL; tokens without `exp` are then not cached at all.
```
Comment (tareknaser) [inline docs/cedarling/reference/cedarling-properties.md:122]: Do you think we should add a CHANGELOG or a release-note entry to mention these updates? We changed some important default values

File: jans-cedarling/cedarling/src/bootstrap_config/jwt_config.rs (hunk 1 above)
Comment (olehbozhok) [reply jans-cedarling/cedarling/src/bootstrap_config/jwt_config.rs:105]: thanks, updated

File: docs/cedarling/reference/cedarling-properties.md (hunk 3 above)
Comment (olehbozhok) [reply docs/cedarling/reference/cedarling-properties.md]: thanks, updated

File: docs/cedarling/reference/cedarling-properties.md (hunk 4 above)
Comment (olehbozhok) [reply docs/cedarling/reference/cedarling-properties.md:122]: I added commit https://github.com/JanssenProject/jans/pull/14098/changes/47eca172008bd19199fa504508daf292108a9f89
it has `fix(jans-cedarling)!` and the comment will appear in the change log

**Comment by olehbozhok** [comment]
@coderabbitai approve

**Review by haileyesus2433 — APPROVED** [review]
